lib/macho/headers.rb

```diff
@@ -815,54 +733,31 @@ def to_h
     # Prelinked kernel/"kernelcache" header structure
     class PrelinkedKernelHeader < MachOStructure
       # @return [Integer] the magic number for a compressed header ({COMPRESSED_MAGIC})
-      attr_reader :signature
+      field :signature, :uint32, :endian => :big
 
       # @return [Integer] the type of compression used
-      attr_reader :compress_type
+      field :compress_type, :uint32, :endian => :big
 
       # @return [Integer] a checksum for the uncompressed data
-      attr_reader :adler32
+      field :adler32, :uint32, :endian => :big
 
       # @return [Integer] the size of the uncompressed data, in bytes
-      attr_reader :uncompressed_size
+      field :uncompressed_size, :uint32, :endian => :big
 
       # @return [Integer] the size of the compressed data, in bytes
-      attr_reader :compressed_size
+      field :compressed_size, :uint32, :endian => :big
 
       # @return [Integer] the version of the prelink format
-      attr_reader :prelink_version
+      field :prelink_version, :uint32, :endian => :big
 
       # @return [void]
-      attr_reader :reserved
+      field :reserved, :binary, :size => 40, :unpack => "L>10"
 
```

Comment:
The name `:binary` might be a little confusing here, since `a{N}` is really a NUL-padded string of up to `{N}` bytes.

Spitballing:

```ruby
field :reserved, :strz, :pad => 40
```

...to emphasize that it's NUL-terminated and that the size is padding.

Reply:
It really is annoying that we don't have one word to describe a null terminated binary slice in English. Originally I had it as `:bin_string` which I think we can both agree is worse and `:binary` seemed like an improvement but I'm not really attached to it or anything. I'll sleep on it and see what I come up with in the morning.

Also, I think that adding `:pad` as an alias for `:size` is a trivial change that would improve the readability of the code a bit. It feels more declarative. Functionally they'd work the same way, right?

Reply:
Okay, after brainstorming a little bit more here are a few mediocre options. Also, I'm not feeling `:pad` anymore. It seems unnecessary imo.

```Ruby
field :reserved, :strz, :size => 40
field :reserved, :var, :size => 40                                              
field :reserved, :varchunk, :size => 40                                         
field :reserved, :varbyte, :size => 40                                          
field :reserved, :nullslice, :size => 40                                        
field :reserved, :nchunk, :size => 40                                           
field :reserved, :nullchunk, :size => 40
```

Reply:
Apparently, `varbyte` already exists as a name to refer to exactly this sort of thing though I doubt it's super well known outside of some database versions. But if we decide to use that then maybe it would be nice to use `varchar` for symmetry instead of `string`. Either way it's your call.

Reply:
The second hardest problem in CS strikes again :slightly_smiling_face: 

I also gave this some more thought, and realized that we might want to decompose this the way we did the `_net` format earlier. Something like this, perhaps:

```ruby
field :foo, :string, :padding => :null, :size => 40
```

or even just have `:padding` be the literal padding byte itself, e.g. `:padding => "\x00"` (although this is harder to convert back into a `#pack`-style format).

This has a few advantages:

* We're separating parsing and semantics: the thing we're extracting is fundamentally a string, not a "strz" or anything else. Ruby calls it a string, so we should call it a string too.
* It gives us the ability to handle alternate paddings, like space paddings. We don't currently need these in Mach-O (AFAIK), but they'd fit naturally in the future.

Thoughts?

Reply:
Kind of unrelated but I just realized that the unpacking directive that's currently used for binary strings doesn't remove trailing nulls and spaces. It just grabs the whole chunk. That's the difference between 'a' (what we're currently using) and 'A'. How confusing!

Following your lead, would we just assume that the basic `:string` type without any options has no padding? I don't hate this idea and it does leave room in the future for other types of padding (though the `String#unpack` method doesn't really have that many options). I prefer `:null` over `\x00` though.
```ruby
field :bstring, :string, :size => 40
field :cstring, :string, :padding => :null, :size => 40
```